initial pass: users endpoint in oc_erchef

marcparadise opened the pull request:
- includes users endpoint support in erchef
- support in underyling components for oc_chef_wm requests
  that do not have a valid org, without stubbing out org.
- make oc_erchef aware of global containers in couch
- nginx routing for endpoints

/ping @opscode/server-team

Note that I'm leaning towards maintaining this as a feature branch
pending completion of other works in progress.

marcparadise committed:
Bring chef_user in line with EC requirements

* add pubkey_version and serialized_object to user record
* allow chef_user to have username specified as ``name`` or ``username``
  for consistency with legacy EC
* chef_user is a chef_object
* simplify validation and ejson construction.  use whitelist of
  accepted fields for serialization
* make display_name and email required and add email validation
* perform proper merge of serialized_object on update .
* silently  drop any fields submitted that we don't recognize
* chef_user:assemble_ejson should only ever include public key,
  not certifcate.
* capture but ignore the ``admin`` and external-auth fields.
* make chef-user accept the standardized form of ``update_from_ejson``,
  which does not include an additional password tuple.  Caller is
  responsible for making sure object record is correct.

sdelano commented on src/chef_user.erl:
```diff
@@ -96,27 +99,54 @@ org_id(#chef_user{}) ->
+    end.
+
 -spec new_record(object_id(), object_id(), {ejson_term(), {binary(), binary(), binary()}}) -> #chef_user{}.
     %% This only works for Open Source Users currently
```
This comment should probably go away

sdelano commented on src/chef_user.erl:
```diff
@@ -96,27 +99,54 @@ org_id(#chef_user{}) ->
     Id = chef_object_base:make_org_prefix_id(OrgId, Name),
     Email = value_or_null({<<"email">>}, UserData),
+    ExtAuthUid = value_or_null({<<"external_authentication_uid">>}, UserData),
+    EnableRecovery = ej:get({"<<recovery_authentication_enabled"}, UserData) =:= true,
```
Is the intent of this line to convert `undefined` to false?

marcparadise replied:
Yes, with a) a typo (badly formatted binary) and b) intent to make it actually do something useful when implementing the associated endpoint.

sdelano commented on src/chef_user.erl:
```diff
@@ -96,27 +99,54 @@ org_id(#chef_user{}) ->
+        end, UserData, [<<"pubkey_version">>, <<"public_key">>, <<"private_key">>, <<"certificate">>,
+                        <<"email">>, <<"admin">>, <<"external_authentication_uid">>,
+                        <<"recovery_authentication_enabled">>, <<"hashed_password">>, <<"salt">>,
+                        <<"username">>, <<"hash_type">>, <<"password">>]).
```
I'd have to doublecheck on exactly what the `opscode-account` code does with user records, but this blacklist will allow users to store **ANYTHING** they'd like in the user record. I don't find this to be particularly friendly to our API. Perhaps, instead, we should implement a whitelist of data that can be stored in the serialized object.

marcparadise replied:
Sounds good to me, I'd been having similar thoughts.  We have some legacy fields that we keep around in json (social media fields that I don't _think_ we're collecting any more - need to check community site though.) 

Similarly,  I'd been considering taking it one step further and removing the serialized_object field altogether, adding the relatively few number of additional fields we have to the schema.  Because of required data migration, though, I shelved it for now.

sdelano commented on src/chef_user.erl:
```diff
@@ -126,60 +156,98 @@ value_or_null(Key, Data) ->
+    {<<"password">>, {fun_match, {fun valid_password/1, string,
+                                  <<"Password must have at least 6 characters">>}}},
+    {<<"display_name">>, string},
+    {<<"email">>, {fun_match,    {fun valid_email/1, string,
```
I see a **TODO** above for `:unless => :external_authentication_enabled`

Still looks like **TODO**

marcparadise replied:
Yes, pending external auth migration to erchef. Will clean up the related TODOs though, it's tracked elsewhere. Edit: will still capture the field, as it is not used in HEC so it can be deployed.  Next round of updates will add the additional validations (and access checks mentioned elsewhere)  as needed.

sdelano commented on src/chef_user.erl:
```diff
@@ -126,60 +156,98 @@ value_or_null(Key, Data) ->
+    {{opt,<<"first_name">>}, string},
+    {{opt,<<"last_name">>}, string},
+    {{opt,<<"password">>}, {fun_match, {fun valid_password/1, string,
+                                        <<"Password must have at least 6 characters">>}}},
```
Again, this should switch on external authentication. It would be nice if we could share the shared validation and be explicit about what validation we do only for create or update. Looks like a bit too much copy-pasta here.

marcparadise replied:
:+1:

sdelano commented on src/chef_user.erl:
```diff
@@ -126,60 +156,98 @@ value_or_null(Key, Data) ->
-                               public_key = PubKey,
-                               admin = Admin},
+                               public_key = KeyOrCert,
+                               admin = _Admin,
```
Do we need this?

sdelano commented on src/chef_user.erl:
```diff
@@ -126,60 +156,98 @@ value_or_null(Key, Data) ->
+    % are not returned in this currently b/c they're not used by opscode-account
+    {[{<<"username">>, Name},
+      {<<"public_key">>, RealPubKey},
+      % TODO not present in OPC, present in OSC - without this, OSC tests will fail.
```
We can leave this on a branch and let the OSC tests fail.

marcparadise replied:
Changes in progress to remove admin throughout allthethings.

sdelano commented on src/chef_user.erl:
```diff
@@ -126,60 +156,98 @@ value_or_null(Key, Data) ->
+    % we don't want to hand the cert back on user GET.
+    RealPubKey = real_pub_key(KeyOrCert),
+    % TODO there are other fields - social media, etc - that
+    % are not returned in this currently b/c they're not used by opscode-account
```
Aren't those fields in the `serialized_object`? It seems like they should be returned by this.

sdelano commented on src/chef_user.erl:
```diff
@@ -189,6 +257,8 @@ parse_binary_json(Bin) ->
 -spec parse_binary_json(binary(), create | update) -> {ok, ej:json_object()}. % or throw
 parse_binary_json(Bin, Operation) ->
+    % TODO we're going to have to move some tsuff out of the blob and into the
+    % serialized_object field of the record?
```
Still a **TODO**? I'm unclear on what you mean here.

marcparadise replied:
stale TODO. Will clean.

sdelano commented on src/chef_user.erl:
```diff
@@ -232,23 +324,50 @@ set_password_data(#chef_user{}=User, {HashedPassword, Salt, HashType}) ->
+update_from_ejson(#chef_user{} = User, {UserEJson, PasswordData}) ->
+    Name = username_from_ejson(UserEJson),
+    Email = ej:get({<<"email">>}, UserEJson),
+    % TODO we'll need to make sure to update this in appropriate places.
```
Still at **TODO**?

sdelano commented on src/chef_user.erl:
```diff
@@ -232,23 +324,50 @@ set_password_data(#chef_user{}=User, {HashedPassword, Salt, HashType}) ->
+    % TODO we'll need to make sure to update this in appropriate places.
+    ExternalAuthenticationUid = ej:get({<<"external_authentication_uid">>}, UserEJson),
+    RecoveryAuthenticationEnabled = ej:get({<<"recovery_authentication_enabled">>}, UserEJson),
+    % TODO  IF we retain this, permissions checks...
```
Don't retain it

sdelano commented on src/chef_user.erl:
```diff
@@ -232,23 +324,50 @@ set_password_data(#chef_user{}=User, {HashedPassword, Salt, HashType}) ->
+    SerializedObject1 = merge_user_data(chef_json:decode(User#chef_user.serialized_object),
+                                        SerializedObject0),
+    User1#chef_user{username = Name,
+                    admin = IsAdmin,
```
No more admin

sdelano commented on src/chef_user.erl:
```diff
@@ -232,23 +324,50 @@ set_password_data(#chef_user{}=User, {HashedPassword, Salt, HashType}) ->
+                    recovery_authentication_enabled = RecoveryAuthenticationEnabled,
+                    serialized_object = SerializedObject1}.
+
+merge_user_data(User, {ModData} = _Data ) ->
```
Why match `_Data`?

sdelano commented on src/chef_user.erl:
```diff
@@ -284,15 +403,26 @@ is_indexed() ->
 ejson_for_indexing(#chef_user{}, _) ->
     error(not_indexed).
 
+% TO password recovery enabled?
```
?

sdelano commented on src/chef_user.erl:
```diff
@@ -284,15 +403,26 @@ is_indexed() ->
+                             username        = UserName,
                              id              = Id }) ->
-    [IsAdmin =:= true, PublicKey, HashedPassword, Salt, HashType, LastUpdatedBy, UpdatedAt, Id].
+    % TODO update username not originally supported in osc, but is in opc. Are any implications?
```
We can update the username in EC? Ugh...

sdelano replied:
I tried to repro in preprod, but it's down. I'd like to verify tomorrow.

sdelano commented:
Also note that this build is failing Travis

marcparadise commented:
Travis build will be fixed with test updates in next push

marcparadise committed:
use a consistently named "name" field when we report error in username

marcparadise committed:
increase coverage in chef_users and ensure all tests pass

* add new tests for additional validation rules
* remove admin-related tests
* make some helper functions for common object creation in tests
* ensure that the validation errors are what we think they are
  by specifying field name in error as part of #ej_invalid{}

Still room to improve, coverage at 67% from ~50.

sdelano commented on src/chef_user.erl:
```diff
@@ -92,124 +84,158 @@ id(#chef_user{id = Id}) ->
+                                              <<"Public Key must be a valid key.">>}}},
+        {{opt,<<"first_name">>},  string},  %% Note that remaining fields are serialized via serialized_object and
+        {{opt,<<"last_name">>},   string} , %% are/were used by other Chef components.
+        {{opt,<<"middle_name">>}, string},  %% FIXME these should be retained by the components that need them
```
What's with the **FIXME** comments here?

marcparadise replied:
Thought I cleared all those in the rebase&squash.  This is another item for the backlog - makes sense to me that required fields should be part of the schema.

sdelano commented on src/chef_user.erl:
```diff
@@ -92,124 +84,158 @@ id(#chef_user{id = Id}) ->
 parse_binary_json(Bin, Operation) ->
   User = chef_object_base:delete_null_public_key(chef_json:decode(Bin)),
   %% If user is invalid, an error is thown
+  validate_user(User, user_spec(common)),
```
Since we handle ej validation errors by catching it's throws, I've got a question about how this works in practice.

Does ej validate the whole spec before returning a list of all the failures, or does it simply fail fast?

If it validates the whole spec, by validating the common components first, we could get ourselves into a situation where we return incomplete validation to a user, they fix it, and still return an invalid object because not all the validation ran the first time.

I'm not really sure how much I'm worried about this in practice.

marcparadise replied:
ej will fail fast, and that bubbles up as a 400 with the associated text.  Agreed it would be better to check them all at once, though I think that's something to add to the list.
